infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/pvc/PVCJobExecutor.java

```diff
@@ -0,0 +1,240 @@
+import org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProjectFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Helps to execute commands needed for workspace PVC preparation and cleanup.
+ *
+ * <p>Creates a short-lived Pod based on CentOS image which mounts a specified PVC and executes a
+ * command (either {@code mkdir -p <path>} or {@code rm -rf <path>}). Reports back whether the pod
+ * succeeded or failed. Supports multiple paths for one command.
+ *
+ * <p>Note that the commands execution is needed only for {@link CommonPVCStrategy}.
+ *
+ * @author amisevsk
+ * @author Anton Korneta
+ */
+@Singleton
+public class PVCJobExecutor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PVCJobExecutor.class);
+
+  static final JobFinishedPredicate POD_PREDICATE = new JobFinishedPredicate();
+
+  static final int COUNT_THREADS = 4;
+  static final int WAIT_POD_TIMEOUT_MIN = 5;
+
+  static final String IMAGE_PULL_POLICY = "IfNotPresent";
+  static final String POD_RESTART_POLICY = "Never";
+  static final String POD_PHASE_SUCCEEDED = "Succeeded";
+  static final String POD_PHASE_FAILED = "Failed";
+  static final String JOB_MOUNT_PATH = "/tmp/job_mount";
+
+  private final String pvcName;
+  private final String jobImage;
+  private final String jobMemoryLimit;
+  private final OpenShiftProjectFactory factory;
+  private final ExecutorService executor;
+
+  @Inject
+  PVCJobExecutor(
+      @Named("che.infra.openshift.pvc.name") String pvcName,
+      @Named("che.openshift.jobs.memorylimit") String jobMemoryLimit,
+      @Named("che.openshift.jobs.image") String jobImage,
+      OpenShiftProjectFactory factory) {
+    this.pvcName = pvcName;
+    this.jobMemoryLimit = jobMemoryLimit;
+    this.jobImage = jobImage;
+    this.factory = factory;
+    this.executor =
+        Executors.newFixedThreadPool(
+            COUNT_THREADS,
+            new ThreadFactoryBuilder()
+                .setNameFormat("PVCJobExecutor-ThreadPool-%d")
+                .setUncaughtExceptionHandler(LoggingUncaughtExceptionHandler.getInstance())
+                .setDaemon(false)
+                .build());
+  }
+
+  /**
+   * Executes the job with the specified arguments.
+   *
+   * @param job the PVC job to execute
+   * @param jobArguments the list of arguments for the specified job
+   * @see PVCJob
+   */
+  void execute(String workspaceId, PVCJob job, String... jobArguments) {
+    final String podName = job.name + '_' + workspaceId;
+    final String[] command = buildCommand(job.baseCommand, jobArguments);
+    final Pod pod = newPod(podName, command);
+    OpenShiftPods pods = null;
+    try {
+      pods = factory.create(workspaceId).pods();
+      pods.create(pod);
+      final Pod finished = pods.wait(podName, WAIT_POD_TIMEOUT_MIN, POD_PREDICATE::apply);
+      if (POD_PHASE_FAILED.equals(finished.getStatus().getPhase())) {
+        throw new InfrastructureException(
```

Comment:
MAybe InternalINfrastructureException is better here?

Reply:
replaced by logging of error